test/polygon_test.cpp

```diff
@@ -131,6 +131,474 @@ TEST(Polygon, MultiMerge) {
+  polys.push_back({
+      {-0.166666672, -0.166666672},  //
+      {-0.462962955, 0.42592594},    //
+      {-0.5, -0.5},                  //
+  });
+  polys.push_back({
+      {-0.240740761, -0.129629627},  //
+      {-0.203703716, -0.092592597},  //
+      {-0.203703716, -0.092592597},  //
+      {-0.203703716, -0.129629627},  //
+  });
+  polys.push_back({
+      {-0.388888896, -0.055555556},  //
+      {-0.277777791, 0.055555556},   //
+      {-0.277777791, -0.055555556},  //
+  });
+  TestPoly(polys, 12);
+};
+
+TEST(Polygon, Holes6) {
+  Polygons polys;
+  polys.push_back({
+      {0.388888896, 0.277777791},  //
+      {0.277777791, 0.388888896},  //
+      {0.388888896, 0.388888896},  //
+  });
+  polys.push_back({
+      {0.055555556, 0.388888896},   //
+      {-0.055555556, 0.277777791},  //
+      {-0.055555556, 0.388888896},  //
+  });
+  polys.push_back({
+      {-0.277777791, 0.388888896},  //
+      {-0.388888896, 0.277777791},  //
+      {-0.388888896, 0.388888896},  //
+  });
+  polys.push_back({
+      {-0.5, -0.5},  //
+      {0.5, -0.5},   //
+      {0.5, 0.5},    //
+      {-0.5, 0.5},   //
+  });
+  TestPoly(polys, 17);
+};
+
+TEST(Polygon, Holes7) {
+  Polygons polys;
+  polys.push_back({
+      {0.203703731, -0.203703716},  //
+      {-0.388888896, 0.277777791},  //
+      {-0.5, -0.5},                 //
+  });
+  polys.push_back({
+      {-0.129629627, -0.203703716},  //
+      {-0.092592597, -0.203703716},  //
+      {-0.092592597, -0.240740761},  //
+  });
+  polys.push_back({
+      {-0.351851851, -0.203703716},  //
+      {-0.314814836, -0.203703716},  //
+      {-0.314814836, -0.240740761},  //
+  });
+  TestPoly(polys, 11);
+};
+
+TEST(Polygon, Holes8) {
+  Polygons polys;
+  polys.push_back({
+      {-0.166666672, 0.166666672},  //
+      {-0.42592594, 0.42592594},    //
+      {-0.5, -0.5},                 //
+  });
+  polys.push_back({
+      {-0.277777791, -0.055555556},  //
+      {-0.388888896, 0.055555556},   //
+      {-0.277777791, 0.055555556},   //
+      {-0.277777791, -0.055555556},  //
+  });
+  polys.push_back({
+      {-0.462962955, -0.129629627},  //
+      {-0.42592594, -0.092592597},   //
+      {-0.42592594, -0.129629627},   //
+  });
+  TestPoly(polys, 12);
+};
+
+TEST(Polygon, Holes9) {
+  Polygons polys;
+  polys.push_back({
+      {-0.166666672, -0.166666672},  //
+      {-0.166666672, 0.166666672},   //
+      {-0.166666657, 0.166666672},   //
+      {-0.462962955, 0.42592594},    //
+      {-0.5, -0.5},                  //
+  });
+  polys.push_back({
+      {-0.277777791, -0.055555556},  //
+      {-0.277777791, -0.055555556},  //
+      {-0.388888896, 0.055555556},   //
+      {-0.277777791, 0.055555556},   //
+  });
+  polys.push_back({
+      {-0.462962955, -0.129629627},  //
+      {-0.42592594, -0.092592597},   //
+      {-0.42592594, -0.092592597},   //
+      {-0.42592594, -0.129629627},   //
+  });
+  TestPoly(polys, 15);
+};
+
+TEST(Polygon, Eberly) {
+  Polygons polys;
```

Comment:
Long time, no see @davideberly! I've extended your earclipping algorithm to handle &epsilon;-valid polygons (polygons that may self-intersect, but for which a perturbation of each vertex by less than &epsilon; exists for which the polygon does not self-intersect). It's taken a lot of work, but I'm pleased with the results. 

In the process I found a small error in your paper. It turns out Mark Spoor was wrong and your original minimize-the-angle-approach is correct. Here is a counter-example polygon for finding a mutually visible vertex.

Reply:
Thanks for letting me know. I am in the process of dealing with another issue for the ear-clipping algorithm and will resolve that soon I hope. I will restore the previous algorithm.